### PR TITLE
perf: don't manipulate an empty value

### PR DESCRIPTION
**Pull request by GalacticHypernova**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

We should not iterate and make O(n) operations on an empty values (strings, arrays), just to receive the same empty value back.

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have added tests (if possible).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by GalacticHypernova** [comment]
Well, the failing tests don't really have anything to do with the PR... I haven't changed anything in the component island URL's.

**Comment by GalacticHypernova** [comment]
@danielroe is this something I accidentally did with one of the changes?

File: packages/nuxt/src/core/nitro.ts [hunk 1]
```diff
@@ -288,16 +288,18 @@ export async function initNitro (nuxt: Nuxt & { _nitro?: Nitro }) {
           createRadixRouter({ routes: routeRules })
         )
         const payloadSuffix = nuxt.options.experimental.renderJsonPayloads ? '/_payload.json' : '/_payload.js'
-        for (const route of nitro._prerenderedRoutes || []) {
-          if (!route.error && route.route.endsWith(payloadSuffix)) {
-            const url = route.route.slice(0, -payloadSuffix.length) || '/'
-            const rules = defu({}, ...routeRulesMatcher.matchAll(url).reverse()) as Record<string, any>
-            if (!rules.prerender) {
-              prerenderedRoutes.add(url)
+        if (nitro._prerenderedRoutes?.length) {
```
Comment (danielroe) [inline packages/nuxt/src/core/nitro.ts:291]: I don't see an issue with the previous code here...

File: packages/nuxt/src/core/plugins/dev-only.ts [hunk 2]
```diff
@@ -21,11 +21,11 @@ export const DevOnlyPlugin = createUnplugin((options: DevOnlyPluginOptions) => {
       if (!DEVONLY_COMP_SINGLE_RE.test(code)) { return }
 
       const s = new MagicString(code)
-      for (const match of code.matchAll(DEVONLY_COMP_RE) || []) {
+      for (const match of code.matchAll(DEVONLY_COMP_RE)) {
         const ast: Node = parse(match[0]).children[0]
         const fallback: Node | undefined = ast.children?.find((n: Node) => n.name === 'template' && Object.values(n.attributes).includes('#fallback'))
         const replacement = fallback ? match[0].slice(fallback.loc[0].end, fallback.loc[fallback.loc.length - 1].start) : ''
-
+
```
Comment (danielroe) [inline packages/nuxt/src/core/plugins/dev-only.ts:28]:
```suggestion

```

File: packages/nuxt/src/core/nitro.ts (hunk 1 above)
Comment (GalacticHypernova) [reply packages/nuxt/src/core/nitro.ts:291]: This is kind of the same principle as everything else, if the array is empty the performance overhead to begin checking the iteration (even if exiting immediately) is much bigger than checking for a truthy length. I'll be able to append benchmark results when I get on PC but there was a notable increase in performance.

File: packages/nuxt/src/core/plugins/dev-only.ts (hunk 2 above)
Comment (GalacticHypernova) [reply packages/nuxt/src/core/plugins/dev-only.ts]: Thanks for the catch!

File: packages/nuxt/src/core/nitro.ts (hunk 1 above)
Comment (GalacticHypernova) [reply packages/nuxt/src/core/nitro.ts:291]: Update: here it is.

```ts
const routeRulesMatcher=""
function defu(table, ...args){

}
const prerenderedRoutes=new Set()
const payloadSuffix=""
const nitro={
    _prerenderedRoutes:[]
}
console.time("current")
for (let index = 0; index < 100000; index++) {

    for (const route of nitro._prerenderedRoutes || []) {
        if (!route.error && route.route.endsWith(payloadSuffix)) {
          const url = route.route.slice(0, -payloadSuffix.length) || '/'
          const rules = defu({}, ...routeRulesMatcher.matchAll(url).reverse())
          if (!rules.prerender) {
            prerenderedRoutes.add(url)
          }
        }
    }
}
console.timeEnd("current")
console.time("PR")
for (let index = 0; index < 100000; index++) {
    if (nitro._prerenderedRoutes?.length) {
        for (const route of nitro._prerenderedRoutes) {
          if (!route.error && route.route.endsWith(payloadSuffix)) {
            const url = route.route.slice(0, -payloadSuffix.length) || '/'
            const rules = defu({}, ...routeRulesMatcher.matchAll(url).reverse())
            if (!rules.prerender) {
              prerenderedRoutes.add(url)
            }
          }
        }
    }
}

console.timeEnd("PR")
```

image

**Review by danielroe — APPROVED** [review]
Thank you ❤️

**Comment by GalacticHypernova** [comment]
Always glad to help!

**Pull request merged**